internal/oc/exporter.go

```diff
@@ -3,41 +3,107 @@ package oc
-	baseEntry := logrus.WithFields(logrus.Fields(s.Attributes))
-	baseEntry.Data["traceID"] = s.TraceID.String()
-	baseEntry.Data["spanID"] = s.SpanID.String()
-	baseEntry.Data["parentSpanID"] = s.ParentSpanID.String()
-	baseEntry.Data["startTime"] = s.StartTime
-	baseEntry.Data["endTime"] = s.EndTime
-	baseEntry.Data["duration"] = s.EndTime.Sub(s.StartTime).String()
-	baseEntry.Data["name"] = s.Name
-	baseEntry.Time = s.StartTime
+	if s.DroppedAnnotationCount > 0 {
+		logrus.WithFields(logrus.Fields{
+			"name":            s.Name,
+			logfields.TraceID: s.TraceID.String(),
+			logfields.SpanID:  s.SpanID.String(),
+			"dropped":         s.DroppedAttributeCount,
+			"maxAttributes":   len(s.Attributes),
+		}).Warning("span had dropped attributes")
+	}
+
+	entry := log.L.Dup()
+	// Combine all span annotations with span data (eg, trace ID, span ID, parent span ID,
+	// error, status code)
+	// (OC) Span attributes are guaranteed to be  strings, bools, or int64s, so we can
+	// can skip overhead in entry.WithFields() and add them directly to entry.Data.
+	// Preallocate ahead of time, since we should add, at most, 10 additional entries
+	data := make(logrus.Fields, len(entry.Data)+len(s.Attributes)+10)
+
+	// Default log entry may have prexisting/application-wide data
+	for k, v := range entry.Data {
+		data[k] = v
+	}
+	for k, v := range s.Attributes {
+		data[k] = v
+	}
+
+	data[logfields.Name] = s.Name
+	data[logfields.TraceID] = s.TraceID.String()
+	data[logfields.SpanID] = s.SpanID.String()
+	data[logfields.ParentSpanID] = s.ParentSpanID.String()
+	data[logfields.StartTime] = s.StartTime
+	data[logfields.EndTime] = s.EndTime
+	data[logfields.Duration] = s.EndTime.Sub(s.StartTime)
+	if sk := spanKindToString(s.SpanKind); sk != "" {
+		data["spanKind"] = sk
+	}
 
 	level := logrus.InfoLevel
 	if s.Status.Code != 0 {
 		level = logrus.ErrorLevel
-		baseEntry.Data[logrus.ErrorKey] = s.Status.Message
+
+		// don't overwrite an existing "error" or "errorCode" attributes
+		if _, ok := data[logrus.ErrorKey]; !ok {
+			data[logrus.ErrorKey] = s.Status.Message
+		}
+		if _, ok := data[_errorCodeKey]; !ok {
+			data[_errorCodeKey] = codes.Code(s.Status.Code).String()
+		}
+	}
+
+	// // if caller information wont be overridden, use that to store span name
+	// if entry.Logger == nil || !entry.Logger.ReportCaller {
+	// 	entry.Caller = &runtime.Frame{
+	// 		Function: s.Name,
+	// 	}
+	// }
+
+	entry.Data = data
+	entry.Time = s.StartTime
+	entry.Log(level, spanMessage)
+}
+
+// todo: after adding GetSpanName option to etwlogrus, switch from Name field in data
+// to overridding entry.Caller information (and ensuring `logger.ReportCaller == false`)
+
+// GetSpanName checks if the entry appears to be span entry exported by the `LogrusExporter`
+// and returns the span name.
+//
+// This is intended to be used with "github.com/Microsoft/go-winio/pkg/etwlogrus".WithGetName.
+func GetSpanName(e *logrus.Entry) string {
```

Comment:
ETW event name is generally used more to indicate the type of event (e.g. the schema). I don't think putting the name of the span in there is a great idea. It would make more sense to make it something like `SpanEvent` instead IMO.

Reply:
Im going to remove this, since I never got around to using it and it doesnt fit, but fwiw, I based it off the Microsoft ETW OpenTelemetry exporter, which does set the span name as the ETW event name:
https://github.com/open-telemetry/opentelemetry-cpp/blob/b04676454d505853cdebf2744412c047de5dd82d/exporters/etw/include/opentelemetry/exporters/etw/etw_tracer.h#L521
https://github.com/open-telemetry/opentelemetry-cpp/blob/d72dcb69b07ffe773e490ee797126357480b9cb0/exporters/etw/include/opentelemetry/exporters/etw/etw_provider.h#L451-L468

Reply:
Thanks for the link, that was an interesting read... They appear to be doing something interesting. They aren't actually sending valid TraceLogging, or manifested ETW, or WPP. They are just using ETW as an envelope, and the data inside is actually MessagePack-encoded JSON:

https://github.com/open-telemetry/opentelemetry-cpp/blob/15ab9ed9adb8150ff1f487a7e31303d6c12e863e/exporters/etw/include/opentelemetry/exporters/etw/etw_provider.h#L370-L382

So, it's different for them, since the event name they are setting has nothing to do with TraceLogging event name, which is what we set. I do think we should keep to the usual semantics of TraceLogging event name, though sounds like it's a moot point for now. :)